writing/2019-10-05-Designing-testable-components.md

```diff
@@ -0,0 +1,471 @@
+
+Let's solve this by parameterizing the function - with another function!
+
+```haskell
+insertNewUser :: Database -> UserName -> Password -> IO UserId
+insertNewUser db user pass =
+  let
+    insertSql =
+      "INSERT INTO table (user_name, password) VALUES (?, ?) returning id"
+  in
+    query db insertSql (user, pass)
+
+createNewUser ::
+     (UserName -> Password -> IO UserId)
+  -> RequestBody
+  -> IO (Either Error User)
+createNewUser persistUser body =
+  case bodyToUser body of
+    Left err -> pure . Left $ err
+    Right (user, pass) -> do
+      -- Persist user:
+      userId <- persistUser user pass
+
+      -- Create a response from the persisted argument:
+      pure . Right $ User { userName = user, userId = userId }
+```
+
+From an approachability standpoint, it is now quite easy to understand how this
+works and how to use the function.
+
+When calling `createNewUser` we now simply provide the `insertNewUser` function
+partially applied with a database.
+
+## Solving the problem at scale
+So here's the problem with the above solution: while it does work, it doesn't
+really scale. Domain logic will often need access to several interfaces to do
+its job. It might need both an HTTP client for some request and a database to
+store the result. As the requirements grow, the solution above quickly becomes
+quite verbose in practice.
+
+E.g:
+
+```haskell
+validateUser :: (UserId -> IO (Maybe User)) -> (UserName -> IO Bool) -> UserId -> IO Bool
+validateUser getUser validateUsername userId = do
+  userM <- getUser userId
+  maybe (pure False) (validateUsername . userName) userM
+```
+
+This example adds one function as argument, but what if you add a third? A
+fourth? You get the picture.
+
+We also don't want to write our code in `IO` - while useful of course, the
+surface area of possible effects is huge. From an effect perspective, we'd like
+to limit the power of each component. We'll get to this later in the article,
+but first let's focus on solving scalability of this initial approach.
+
+### Introducing the Handle pattern
+Instead of parameterizing the function with another function we can
+parameterize with a datatype containing a function. This is great for a number
+of reasons.
+
+1. We can group functions that operate in similar ways together (think OO
+   interface)
+2. We don't have to pass around all those functions, instead we pass one
+   datatype
+
+So how would this look?
+
+```haskell
+data Application =
+  { persistUser :: UserName -> Password -> IO UserId
+  , getUser :: UserId -> IO (Maybe User)
+  , validateUsername :: UserId -> IO Bool
+  , logLn :: Loggable a => a -> IO ()
+  }
+```
+
+Now we can simply pass that to `validateUser`:
+
+```haskell
+validateUser :: Application -> RequestBody -> UserId -> IO Bool
+validateUser app requestBody userId = do
+  userM <- (app & getUser) userId
+  maybe (pure False) (app & validateUsername $ userName) userM
```

Comment:
An alternative way to write this code without using `&` is to use something like `RecordWildCards` or `NamedFieldPuns`. Since method names are not prefixed with `app`, the usage of `RecordWildCards` can be confusing. But `NamedFieldPuns` should do the trick if you don't mind some verbosity. I also think that for beginners such code is more readable than using operator `&`.

```haskell
validateUser :: Application -> RequestBody -> UserId -> IO Bool
validateUser App{getUser, validateUsername} requestBody userId = do
    userM <- getUser userId
    maybe (pure False) (validateUsername userName) userM
```